babel/src/board.js

```diff
@@ -283,20 +280,98 @@ export class Board {
         let hasWideStyleMatch = _.some(_.map(flatChunks, hasMatchInSingleRow));
         return hasWideStyleMatch || _.some(_.map(chunks), hasMatchInPairOfRows);
     }
+    
+    hasMatchEvent() {
+        if (findMatches(this.orbs)[0]) { return true };
+    }
 
-    swap(swapOrbs) {
-        let [[x1, y1], [x2, y2]] = swapOrbs;
+    swap(swapOrbs, playerSwap = true) {
+        let [[row1, col1], [row2, col2]] = swapOrbs;
         let orbsBefore = _.cloneDeep(this.orbs);
-        this.orbs[x1][y1] = orbsBefore[x2][y2]
-        this.orbs[x2][y2] = orbsBefore[x1][y1]
+        this.orbs[row1][col1] = orbsBefore[row2][col2]
+        this.orbs[row2][col2] = orbsBefore[row1][col1]
 
         // undo the swap if it did not yeild a match
-        if (!findMatches(this.orbs)[0]) {
+        if (playerSwap && !this.hasMatchEvent()) {
             this.orbs = orbsBefore;
         };
     }
-
+    
     shuffle() {
+        // shuffle the board once here
+        while (this.hasMatchEvent() || this.needsShuffle()) {
+            this._unMatch();
+        }
+    }
 
+    _safeToSwap(row, col, dir) {
+        let neighbors = [];
+        if (dir === 'up' && !_.isUndefined(this.orbs[row - 1])) {
+            if (_.isUndefined(this.orbs[row - 3])) {
+                return true
+            } else {
+                neighbors = [this.orbs[row - 2][col], this.orbs[row - 3][col]]
+            }
+        } else if (dir === 'down' && !_.isUndefined(this.orbs[row + 1])) {
+            if (_.isUndefined(this.orbs[row + 3])) {
+                return true;
+            } else {
+                neighbors = [this.orbs[row + 2][col], this.orbs[row + 3][col]];
+            }
+        } else if (dir === 'left' && !_.isUndefined(this.orbs[row][col - 1])) {
+            if (_.isUndefined(this.orbs[row][col - 3])) {
+                return true
+            } else {
+                neighbors = [this.orbs[row][col - 2], this.orbs[row][col - 3]]
+            }
+        } else if (dir === 'right' && !_.isUndefined(this.orbs[row][col + 1])) {
+            if (_.isUndefined(this.orbs[row][col + 3])) {
+                return true
+            } else {
+                neighbors = [this.orbs[row][col + 2], this.orbs[row][col + 3]]
+            }
+        } else {
+            return false
+        };
+        return !_.includes(neighbors, this.orbs[row][col]);
+    }
+
+    _replaceWithDifferentOrb(row, col) {
+        this.orbs[row][col] = _.sample(_.difference(this.types, this.orbs[row][col]));
+    }
+
+    _unMatch() {
+        let match = combineMatches(findMatches(this.orbs))[0];
+        let [firstRow, firstCol] = match[0];
+        let [secondRow, secondCol] = match[1];
+        let [lastRow, lastCol] = _.last(match);
+        let isHorizontal = (firstRow === secondRow);
+        let isVertical = (firstCol === secondCol);
```

Comment:
A cross match with coords such as `[[[1, 2], [2, 1], [2, 2], [2, 3], [3, 2]]]` does not show as either horizontal or vertical, so it defaults to replacing the third orb when it doesn't have to.